File: bot-api/java/src/main/java/dev/robocode/tankroyale/botapi/internal/WebSocketHandler.java
package dev.robocode.tankroyale.botapi.internal;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import dev.robocode.tankroyale.botapi.*;
import dev.robocode.tankroyale.botapi.events.ConnectedEvent;
import dev.robocode.tankroyale.botapi.events.ConnectionErrorEvent;
import dev.robocode.tankroyale.botapi.events.DisconnectedEvent;
import dev.robocode.tankroyale.botapi.events.GameEndedEvent;
import dev.robocode.tankroyale.botapi.events.GameStartedEvent;
import dev.robocode.tankroyale.botapi.events.RoundEndedEvent;
import dev.robocode.tankroyale.botapi.events.RoundStartedEvent;
import dev.robocode.tankroyale.botapi.events.SkippedTurnEvent;
import dev.robocode.tankroyale.botapi.internal.json.JsonConverter;
import dev.robocode.tankroyale.botapi.mapper.EventMapper;
import dev.robocode.tankroyale.botapi.mapper.GameSetupMapper;
import dev.robocode.tankroyale.schema.BotReady;
import dev.robocode.tankroyale.schema.GameEndedEventForBot;
import dev.robocode.tankroyale.schema.GameStartedEventForBot;
import dev.robocode.tankroyale.schema.Message;
import dev.robocode.tankroyale.schema.TickEventForBot;
import dev.robocode.tankroyale.schema.ServerHandshake;

import java.net.URI;
import java.net.http.WebSocket;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.CountDownLatch;

import static dev.robocode.tankroyale.botapi.mapper.ResultsMapper.map;

final class WebSocketHandler implements WebSocket.Listener {

    private final BaseBotInternals baseBotInternals;
    private final URI serverUrl;
    private final String serverSecret;
    private final IBaseBot baseBot;
    private final BotInfo botInfo;
    private final BotEventHandlers botEventHandlers;
    private final InternalEventHandlers internalEventHandlers;
    private final CountDownLatch closedLatch;

    private WebSocket socket;
    private final StringBuilder payload = new StringBuilder();

    WebSocketHandler(
            BaseBotInternals baseBotInternals,
            URI serverUrl,
            String serverSecret,
            IBaseBot baseBot,
            BotInfo botInfo,
            BotEventHandlers botEventHandlers,
            InternalEventHandlers internalEventHandlers,
            CountDownLatch closedLatch) {
        this.baseBotInternals = baseBotInternals;
        this.serverUrl = serverUrl;
        this.serverSecret = serverSecret;
        this.baseBot = baseBot;
        this.botInfo = botInfo;
        this.botEventHandlers = botEventHandlers;
        this.internalEventHandlers = internalEventHandlers;
        this.closedLatch = closedLatch;
    }

    @Override
    public void onOpen(WebSocket websocket) {
        this.socket = websocket; // To prevent null pointer exception

        botEventHandlers.onConnected.publish(new ConnectedEvent(serverUrl));
        WebSocket.Listener.super.onOpen(websocket);
    }

    @Override
    public CompletionStage<?> onClose(WebSocket websocket, int statusCode, String reason) {
        var disconnectedEvent = new DisconnectedEvent(serverUrl, true, statusCode, reason);

        botEventHandlers.onDisconnected.publish(disconnectedEvent);
        internalEventHandlers.onDisconnected.publish(disconnectedEvent);

        closedLatch.countDown();
        return null;
    }

    @Override
    public void onError(WebSocket websocket, Throwable error) {
        botEventHandlers.onConnectionError.publish(new ConnectionErrorEvent(serverUrl, error));

        closedLatch.countDown();
    }

    @Override
    public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
        payload.append(data);
        if (last) {
            JsonObject jsonMsg = JsonConverter.fromJson(payload.toString(), JsonObject.class);
            payload.delete(0, payload.length()); // clear payload buffer

            JsonElement jsonType = jsonMsg.get("type");
            if (jsonType != null) {
                String type = jsonType.getAsString();

                switch (dev.robocode.tankroyale.schema.Message.Type.fromValue(type)) {
                    case TICK_EVENT_FOR_BOT:
                        handleTick(jsonMsg);
                        break;
                    case ROUND_STARTED_EVENT:
                        handleRoundStarted(jsonMsg);
                        break;
                    case ROUND_ENDED_EVENT_FOR_BOT:
                        handleRoundEnded(jsonMsg);
                        break;
                    case GAME_STARTED_EVENT_FOR_BOT:
                        handleGameStarted(jsonMsg);
                        break;
                    case GAME_ENDED_EVENT_FOR_BOT:
                        handleGameEnded(jsonMsg);
                        break;
                    case SKIPPED_TURN_EVENT:
                        handleSkippedTurn(jsonMsg);
                        break;
                    case SERVER_HANDSHAKE:
                        handleServerHandshake(jsonMsg);
                        break;
                    case GAME_ABORTED_EVENT:
                        handleGameAborted();
                        break;
                    default:
                        throw new BotException("Unsupported WebSocket message type: " + type);
                }
            }
        }
        return WebSocket.Listener.super.onText(webSocket, data, last);
    }

    private void handleTick(JsonObject jsonMsg) {
        if (baseBotInternals.getEventHandlingDisabledTurn()) return;

        baseBotInternals.setTickStartNanoTime(System.nanoTime());

        var tickEventForBot = JsonConverter.fromJson(jsonMsg, TickEventForBot.class);

        var mappedTickEvent = EventMapper.map(tickEventForBot, baseBot);

        baseBotInternals.addEventsFromTick(mappedTickEvent);

        if (baseBotInternals.getBotIntent().getRescan() != null && baseBotInternals.getBotIntent().getRescan()) {
            baseBotInternals.getBotIntent().setRescan(false);
        }

        baseBotInternals.setTickEvent(mappedTickEvent);

        mappedTickEvent.getEvents().forEach(internalEventHandlers::fireEvent);

        // Trigger next turn (not tick-event!)
        internalEventHandlers.onNextTurn.publish(mappedTickEvent);
    }

    private void handleRoundStarted(JsonObject jsonMsg) {
        var roundStartedEvent = JsonConverter.fromJson(jsonMsg, RoundStartedEvent.class);

        var mappedRoundStartedEvent = new RoundStartedEvent(roundStartedEvent.getRoundNumber());

        botEventHandlers.onRoundStarted.publish(mappedRoundStartedEvent);
        internalEventHandlers.onRoundStarted.publish(mappedRoundStartedEvent);
    }

    private void handleRoundEnded(JsonObject jsonMsg) {
        var roundEndedEvent = JsonConverter.fromJson(jsonMsg, RoundEndedEvent.class);

        var mappedRoundEndedEvent = new RoundEndedEvent(
                roundEndedEvent.getRoundNumber(), roundEndedEvent.getTurnNumber(), roundEndedEvent.getResults());

        botEventHandlers.onRoundEnded.publish(mappedRoundEndedEvent);
        internalEventHandlers.onRoundEnded.publish(mappedRoundEndedEvent); // triggers stopThread()

        // Dispatch any queued events (e.g. WonRoundEvent from the last tick). Bot thread is now
        // stopped so there is no concurrent dispatch race. Must run before ROUND_STARTED clears
        // the event queue.
        baseBotInternals.dispatchEvents(mappedRoundEndedEvent.getTurnNumber());

        // Transfer any remaining stdout/stderr from event handlers (e.g. onWonRound) before the round ends
        baseBotInternals.transferStdOutToBotIntent();
    }

    private void handleGameStarted(JsonObject jsonMsg) {
        var gameStartedEventForBot = JsonConverter.fromJson(jsonMsg, GameStartedEventForBot.class);

        baseBotInternals.setMyId(gameStartedEventForBot.getMyId());

        Set<Integer> teammateIds = gameStartedEventForBot.getTeammateIds() == null ?
                Set.of() : new HashSet<>(gameStartedEventForBot.getTeammateIds());
        baseBotInternals.setTeammateIds(teammateIds);

        baseBotInternals.setGameSetup(GameSetupMapper.map(gameStartedEventForBot.getGameSetup()));

        InitialPosition initialPosition = new InitialPosition(
                gameStartedEventForBot.getStartX(),
                gameStartedEventForBot.getStartY(),
                gameStartedEventForBot.getStartDirection());
        baseBotInternals.setInitialPosition(initialPosition);

        // Send ready signal
        var ready = new BotReady();
        ready.setType(Message.Type.BOT_READY);

        String msg = JsonConverter.toJson(ready);
        socket.sendText(msg, true);

        botEventHandlers.onGameStarted.publish(
                new GameStartedEvent(gameStartedEventForBot.getMyId(), initialPosition, baseBotInternals.getGameSetup()));
    }

    private void handleGameEnded(JsonObject jsonMsg) {
        // Send the game ended event
        var gameEndedEventForBot = JsonConverter.fromJson(jsonMsg, GameEndedEventForBot.class);

        var mappedGameEnded = new GameEndedEvent(
                gameEndedEventForBot.getNumberOfRounds(),
                map(gameEndedEventForBot.getResults()));

        botEventHandlers.onGameEnded.publish(mappedGameEnded);
        internalEventHandlers.onGameEnded.publish(mappedGameEnded);
    }

    private void handleGameAborted() {
        botEventHandlers.onGameAborted.publish(null);
        internalEventHandlers.onGameAborted.publish(null);
    }

    private void handleSkippedTurn(JsonObject jsonMsg) {
        var skippedTurnEvent = JsonConverter.fromJson(jsonMsg, dev.robocode.tankroyale.schema.SkippedTurnEvent.class);

        baseBotInternals.addEvent((SkippedTurnEvent) EventMapper.map(skippedTurnEvent, baseBot));
    }

    private void handleServerHandshake(JsonObject jsonMsg) {
        var serverHandshake = JsonConverter.fromJson(jsonMsg, ServerHandshake.class);
        baseBotInternals.setServerHandshake(serverHandshake);

        // Validate bot info before sending bot handshake
        validateBotInfo();

        // Reply by sending bot handshake
        var isDroid = baseBot instanceof Droid;
        var botHandshake = BotHandshakeFactory.create(serverHandshake.getSessionId(), botInfo, isDroid, serverSecret);
        String msg = JsonConverter.toJson(botHandshake);

        socket.sendText(msg, true);
    }

    private void validateBotInfo() {
        // If the bot is booted, botInfo might be partially filled by the booter
        // but the bot code must ensure name, version, and authors are present.
        if (isBlank(botInfo.getName())) {
            throwMissingPropertyException("name");
        }
        if (isBlank(botInfo.getVersion())) {
            throwMissingPropertyException("version");
        }
        if (botInfo.getAuthors() == null || botInfo.getAuthors().isEmpty() || isAllBlank(botInfo.getAuthors())) {
            throwMissingPropertyException("authors");
        }
    }

    private void throwMissingPropertyException(String propertyName) {
        throw new BotException(
                String.format("Required bot property '%s' is missing. " +
                        "This property is required in order for the bot to be recognized when booting it up and " +
                        "when it needs to join the game. You must set this property in your bot code " +
                        "or provide a .json configuration file.", propertyName));
    }

    private static boolean isBlank(String s) {
        return s == null || s.trim().isEmpty();
    }

    private static boolean isAllBlank(Iterable<String> iterable) {
        for (String s : iterable) {
            if (!isBlank(s)) {
                return false;
            }
        }
        return true;
    }
}

File: bot-api/typescript/src/WebSocketHandler.ts
import { BotInfo } from "./BotInfo.js";
import { BotHandshakeFactory } from "./BotHandshakeFactory.js";
import { EnvVars } from "./EnvVars.js";
import { BotException } from "./BotException.js";
import { RuntimeAdapter } from "./runtime/RuntimeAdapter.js";
import { WebSocketLike } from "./runtime/WebSocketLike.js";
import { MessageType } from "./protocol/MessageType.js";
import type {
  ServerHandshake,
  TickEventForBot,
  GameStartedEventForBot,
  GameEndedEventForBot,
  RoundStartedEvent,
  RoundEndedEventForBot,
  BotIntent,
  BotReady,
} from "./protocol/schema.js";

/** Callback hooks provided by the bot internals layer. */
export interface WebSocketHandlerCallbacks {
  onConnected?: () => void;
  onDisconnected?: (remote: boolean, code?: number, reason?: string) => void;
  onConnectionError?: (error: unknown) => void;
  onServerHandshake?: (handshake: ServerHandshake) => void;
  onGameStarted?: (event: GameStartedEventForBot) => void;
  onGameEnded?: (event: GameEndedEventForBot) => void;
  onGameAborted?: () => void;
  onRoundStarted?: (event: RoundStartedEvent) => void;
  onRoundEnded?: (event: RoundEndedEventForBot) => void;
  onTick?: (event: TickEventForBot) => void;
  onSkippedTurn?: (turnNumber: number) => void;
}

/**
 * Handles the WebSocket connection to the server, routing incoming messages
 * to the appropriate callback hooks. Matches Java's WebSocketHandler behavior.
 */
export class WebSocketHandler {
  private readonly adapter: RuntimeAdapter;
  private readonly serverUrl: string;
  private readonly serverSecret: string | undefined;
  private readonly botInfo: BotInfo;
  private readonly envVars: EnvVars;
  private readonly callbacks: WebSocketHandlerCallbacks;

  private socket: WebSocketLike | null = null;
  private serverHandshake: ServerHandshake | null = null;

  constructor(
    adapter: RuntimeAdapter,
    serverUrl: string,
    serverSecret: string | undefined,
    botInfo: BotInfo,
    envVars: EnvVars,
    callbacks: WebSocketHandlerCallbacks,
  ) {
    this.adapter = adapter;
    this.serverUrl = serverUrl;
    this.serverSecret = serverSecret;
    this.botInfo = botInfo;
    this.envVars = envVars;
    this.callbacks = callbacks;
  }

  /** Connects to the server WebSocket. */
  connect(): void {
    const ws = this.adapter.createWebSocket(this.serverUrl);
    this.socket = ws;

    ws.onopen = () => {
      this.callbacks.onConnected?.();
    };

    ws.onclose = (event) => {
      const e = event as { code?: number; reason?: string };
      this.callbacks.onDisconnected?.(true, e.code, e.reason);
    };

    ws.onerror = (event) => {
      this.callbacks.onConnectionError?.(event);
    };

    ws.onmessage = (event) => {
      this.handleMessage(String(event.data));
    };
  }

  /** Disconnects from the server WebSocket. */
  disconnect(): void {
    this.socket?.close();
    this.socket = null;
  }

  /** Sends a {@link BotIntent} message to the server. */
  sendBotIntent(intent: BotIntent): void {
    this.send(intent);
  }

  /** Returns the stored server handshake, or null if not yet received. */
  getServerHandshake(): ServerHandshake | null {
    return this.serverHandshake;
  }

  // ---------------------------------------------------------------------------
  // Message routing
  // ---------------------------------------------------------------------------

  private handleMessage(json: string): void {
    try {
      let msg: { type?: string };
      try {
        msg = JSON.parse(json) as { type?: string };
      } catch {
        throw new BotException("Failed to parse WebSocket message: " + json);
      }

      const type = msg.type;
      if (type == null) return;

      switch (type as MessageType) {
        case MessageType.ServerHandshake:
          this.handleServerHandshake(msg as unknown as ServerHandshake);
          break;
        case MessageType.TickEventForBot:
          this.handleTick(msg as unknown as TickEventForBot);
          break;
        case MessageType.RoundStartedEvent:
          this.handleRoundStarted(msg as unknown as RoundStartedEvent);
          break;
        case MessageType.RoundEndedEventForBot:
          this.handleRoundEnded(msg as unknown as RoundEndedEventForBot);
          break;
        case MessageType.GameStartedEventForBot:
          this.handleGameStarted(msg as unknown as GameStartedEventForBot);
          break;
        case MessageType.GameEndedEventForBot:
          this.handleGameEnded(msg as unknown as GameEndedEventForBot);
          break;
        case MessageType.SkippedTurnEvent:
          this.handleSkippedTurn(msg as unknown as { turnNumber: number });
          break;
        case MessageType.GameAbortedEvent:
          this.handleGameAborted();
          break;
        default:
          throw new BotException("Unsupported WebSocket message type: " + type);
      }
    } catch (err) {
      this.callbacks.onConnectionError?.(err);
    }
  }

  private handleServerHandshake(msg: ServerHandshake): void {
    this.serverHandshake = msg;
    this.callbacks.onServerHandshake?.(msg);

    // Validate required bot info before sending the bot handshake
    this.validateBotInfo();

    const isDroid = false; // Droid detection handled by caller via BotInfo/callbacks
    const handshake = BotHandshakeFactory.create(
      msg.sessionId,
      this.botInfo,
      isDroid,
      this.serverSecret,
      this.envVars,
    );
    this.send(handshake);
  }

  private handleGameStarted(msg: GameStartedEventForBot): void {
    this.callbacks.onGameStarted?.(msg);

    const ready: BotReady = { type: MessageType.BotReady };
    this.send(ready);
  }

  private handleTick(msg: TickEventForBot): void {
    this.callbacks.onTick?.(msg);
  }

  private handleRoundStarted(msg: RoundStartedEvent): void {
    this.callbacks.onRoundStarted?.(msg);
  }

  private handleRoundEnded(msg: RoundEndedEventForBot): void {
    this.callbacks.onRoundEnded?.(msg);
  }

  private handleGameEnded(msg: GameEndedEventForBot): void {
    this.callbacks.onGameEnded?.(msg);
  }

  private handleGameAborted(): void {
    this.callbacks.onGameAborted?.();
  }

  private handleSkippedTurn(msg: { turnNumber: number }): void {
    this.callbacks.onSkippedTurn?.(msg.turnNumber);
  }

  // ---------------------------------------------------------------------------
  // Helpers
  // ---------------------------------------------------------------------------

  /** Validates that required bot info fields are set before sending the bot handshake. */
  private validateBotInfo(): void {
    if (this.isBlank(this.botInfo.name)) {
      this.throwMissingPropertyException("name");
    }
    if (this.isBlank(this.botInfo.version)) {
      this.throwMissingPropertyException("version");
    }
    const authors = this.botInfo.authors;
    if (authors == null || authors.length === 0 || authors.every((a) => this.isBlank(a))) {
      this.throwMissingPropertyException("authors");
    }
  }

  private throwMissingPropertyException(propertyName: string): never {
    throw new BotException(
      `Required bot property '${propertyName}' is missing. ` +
        `This property is required in order for the bot to be recognized when booting it up and ` +
        `when it needs to join the game. You must set this property in your bot code ` +
        `or provide a .json configuration file.`,
    );
  }

  private isBlank(s: string | null | undefined): boolean {
    return s == null || s.trim() === "";
  }

  private send(obj: object): void {
    if (this.socket == null) {
      throw new BotException("WebSocket is not connected");
    }
    this.socket.send(JSON.stringify(obj));
  }
}
